test/regress_ws.c

```diff
@@ -0,0 +1,362 @@
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include "util-internal.h"
+
+#ifdef _WIN32
+#include <winsock2.h>
+#include <ws2tcpip.h>
+#include <windows.h>
+#endif
+
+#include "event2/event-config.h"
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#ifdef EVENT__HAVE_SYS_TIME_H
+#include <sys/time.h>
+#endif
+#include <sys/queue.h>
+#ifndef _WIN32
+#include <sys/socket.h>
+#include <signal.h>
+#include <unistd.h>
+#include <netdb.h>
+#endif
+#include <string.h>
+
+#include "event2/event.h"
+#include "event2/http.h"
+#include "event2/buffer.h"
+#include "event2/bufferevent_ssl.h"
+#include "event2/ws.h"
+#include "regress.h"
+#include "regress_http.h"
+#include "regress_ws.h"
+
+#define htonll(x)    \
+	((1 == htonl(1)) \
+			? (x)    \
+			: ((uint64_t)htonl((x)&0xFFFFFFFF) << 32) | htonl((x) >> 32))
+#define ntohll(x) htonll(x)
+
+
+static struct event_base *exit_base;
+
+static void
+on_ws_msg_cb(struct evws_connection *evws, char *data, size_t len, void *arg)
+{
+	ev_uintptr_t val = (ev_uintptr_t)arg;
+
+	if (val != 0xDEADBEEF) {
+		fprintf(stdout, "FAILED on_complete_cb argument\n");
+		exit(1);
+	}
+
+	if (!strcmp(data, "Send echo")) {
+		const char *reply = "Reply echo";
+
+		evws_send(evws, reply, strlen(reply));
+		test_ok++;
+	} else if (!strcmp(data, "Client: hello")) {
+		test_ok++;
+	} else if (!strcmp(data, "Close")) {
+		evws_close(evws, 0);
+		test_ok++;
+	} else {
+		/* unexpected test message */
+		event_base_loopexit(arg, NULL);
+	}
+}
+
+static void
+on_ws_close_cb(struct evws_connection *evws, void *arg)
+{
+	ev_uintptr_t val = (ev_uintptr_t)arg;
+
+	if (val != 0xDEADBEEF) {
+		fprintf(stdout, "FAILED on_complete_cb argument\n");
+		exit(1);
+	}
+	test_ok++;
+}
+
+void
+http_on_ws_cb(struct evhttp_request *req, void *arg)
+{
+	struct evws_connection *evws;
+	const char *hello = "Server: hello";
+
+	evws = evws_new_session(req, on_ws_msg_cb, (void *)0xDEADBEEF);
+	if (!evws)
+		return;
+	test_ok++;
+
+	evws_connection_set_closecb(evws, on_ws_close_cb, (void *)0xDEADBEEF);
+	evws_send(evws, hello, strlen(hello));
+}
+
+static void
+http_ws_errorcb(struct bufferevent *bev, short what, void *arg)
+{
+	/** For ssl */
+	if (what & BEV_EVENT_CONNECTED)
+		return;
+	test_ok++;
+	event_base_loopexit(arg, NULL);
+}
+
+static char *
+receive_ws_msg(struct evbuffer *buf, size_t *out_len, unsigned *options)
```

Comment:
And same here, maybe add an internal helper?

Reply:
Didn't get how the sending and receiving routines can share both the same code via helpers.

Reply:
I meant not between sending and receiving, to add a common internal helper and use it in `ws.c` and `regress_ws.c`

Reply:
But in this case it will be bad in case we have common error in both test and implementation

Reply:
I'm fine with leave it as-is, just wanted to reduce amount of code.